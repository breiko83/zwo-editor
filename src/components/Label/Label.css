.label {
  position: absolute;
  z-index: 100;
  top: -115px;
  left: 0;
  color: white;
  background-color: rgba(0, 0, 0, 0.7);
  width: 150px;
  border-radius: 5px;
  margin: 5px 0;
  text-align: center;
  font-size: 14px; 
  padding: 5px;
}

.textField {
  background-color: rgba(255, 255, 255, 0.1);
  color: white;
  border: 0;
  border-bottom: 0px solid white;
  font-size: 14px;
}

.cadence-row {
  display: flex;
}
.cadence-row > .cadenceLabel {
  font-size: 14px;
  flex: 1;
}
.cadence-row > .textField {
  min-width: 1px;
  flex: 1;
}
