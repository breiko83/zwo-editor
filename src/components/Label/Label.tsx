import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBolt, faClock, faRuler } from '@fortawesome/free-solid-svg-icons'
import './Label.css'
import { FreeInterval, RampInterval, SteadyInterval } from '../../types/Interval'
import { WorkoutMode } from '../../modes/WorkoutMode'
import BikeMode from '../../modes/BikeMode'
import RunMode from '../../modes/RunMode'
import * as format from '../../utils/format'

interface LabelProps {
  interval: SteadyInterval | RampInterval | FreeInterval;
  mode: WorkoutMode;
  onCadenceChange: (cadence: number) => void;
}

const Label = ({ mode, ...props }: LabelProps) => {
  return (
    <div className='label'>
      <div>
        <FontAwesomeIcon icon={faClock} fixedWidth /> {format.duration(mode.intervalDuration(props.interval))}
      </div>
      {mode instanceof BikeMode ? <BikeData mode={mode} {...props} /> : <RunData mode={mode} {...props} />}
    </div>
  );
};

function BikeData({ interval, mode, onCadenceChange }: LabelProps & { mode: BikeMode }) {
  return (
    <>
      {interval.type === "steady" &&
        <>
          <div>
            <FontAwesomeIcon icon={faBolt} fixedWidth /> {format.power(mode.power(interval.intensity))}
          </div>
          <div>
            {format.wkg(mode.wkg(interval.intensity))} &middot; {format.percentage(interval.intensity)} FTP
          </div>
        </>
      }
      {interval.type === "ramp" &&
        <>
          <div>
            <FontAwesomeIcon icon={faBolt} fixedWidth /> {format.power(mode.power(interval.startIntensity))} - {format.power(mode.power(interval.endIntensity))}
          </div>
          <div>
            {format.percentage(interval.startIntensity)} - {format.percentage(interval.endIntensity)} FTP
          </div>
        </>
      }
      <div className="cadence-row">
        <label className="cadenceLabel">Cadence</label>
        <CadenceInput cadence={interval.cadence} onCadenceChange={onCadenceChange} />
      </div>
    </>
  );
}

function RunData({ interval, mode }: LabelProps & { mode: RunMode }) {
  return (
    <>
      <div>
        <FontAwesomeIcon icon={faRuler} fixedWidth /> {format.distance(mode.intervalDistance(interval))}
      </div>
      {interval.type === "steady" &&
        <div>
          {format.percentage(interval.intensity)} {format.shortPaceName(interval.pace)}
        </div>
      }
      {interval.type === "ramp" &&
        <div>
          {format.percentage(interval.startIntensity)} to {format.percentage(interval.endIntensity)} {format.shortPaceName(interval.pace)}
        </div>
      }
    </>
  );
}

function CadenceInput({cadence, onCadenceChange}: {cadence: number, onCadenceChange: (v: number) => void}) {
  return (
    <input
      type="number"
      min="40"
      max="150"
      step="5"
      name="cadence"
      value={cadence || ''}
      onChange={(e) => {onCadenceChange(parseInt(e.target.value))}}
      onClick={(e)=> {e.stopPropagation()}}
      className="textField cadence"
    />
  );
}

export default Label