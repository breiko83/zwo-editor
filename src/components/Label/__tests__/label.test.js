import React from 'react';
import Label from '../../Label/Label';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'

test('Label renders correctly', () => {
  const label = {
    duration: 100,
    power: 250,
    weight: 75
  }

  const component = renderer.create(
    <Label duration={label.duration} power={label.power} weight={label.weight} />
  )

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})