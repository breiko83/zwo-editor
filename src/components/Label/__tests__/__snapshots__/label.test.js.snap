// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders correctly 1`] = `
<div
  className="label"
>
  <div>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-clock fa-w-16 fa-fw "
      data-icon="clock"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256,8C119,8,8,119,8,256S119,504,256,504,504,393,504,256,393,8,256,8Zm92.49,313h0l-20,25a16,16,0,0,1-22.49,2.5h0l-67-49.72a40,40,0,0,1-15-31.23V112a16,16,0,0,1,16-16h32a16,16,0,0,1,16,16V256l58,42.5A16,16,0,0,1,348.49,321Z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     
    100
  </div>
  <div
    className="cadence-row"
  >
    <label
      className="cadenceLabel"
    >
      Cadence
    </label>
    <input
      className="textField cadence"
      max="150"
      min="40"
      name="cadence"
      onChange={[Function]}
      onClick={[Function]}
      step="5"
      type="number"
      value=""
    />
  </div>
</div>
`;
